### PR TITLE
Mainly chunk size change

### LINKED ISSUE #1: Ability for specify a chunk size based on size not count
Currently the chunk size is a count of items implementing SortType, but if SortType also had a Size() it could instead use that, which would be very useful! Otherwise great package! 👍

### PR DESCRIPTION
**Pull request by yonderblue**
fixes #1

### CODE REVIEW

File: types.go [hunk 1]
```diff
@@ -4,7 +4,15 @@ import "context"
...
 type SortType interface {
-	ToBytes() []byte // ToBytes used for marshaling with gob
+	ToBytes(buf []byte) []byte
+}
+
+type Sizer interface {
+	Size() int
+}
+
+type Releaser interface {
```
Comment (yonderblue) [inline types.go:14]: Maybe should add a little doc here and on Sizer too.

File: example_test.go [hunk 2]
```diff
@@ -16,8 +16,10 @@ type sortInt struct {
 	i int64
 }
 
-func (s sortInt) ToBytes() []byte {
-	buf := make([]byte, binary.MaxVarintLen64)
+func (s sortInt) ToBytes(buf []byte) []byte {
+	for i := 0; i < binary.MaxVarintLen64; i++ {
```
Comment (lanrat) [inline example_test.go:20]: Why change this?
This seems like a slower way to accomplish the same test.

File: sort_sorttype.go [hunk 3]
```diff
@@ -143,14 +142,26 @@ func (s *SortTypeSorter) buildChunks() error {
 	defer close(s.chunkChan) // if this is not called on error, causes a deadlock
 
 	for {
-		c := newChunk(s.config.ChunkSize, s.lessFunc)
-		for i := 0; i < s.config.ChunkSize; i++ {
+		c := newChunk(s.lessFunc)
+		var size int
+	Loop:
```
Comment (lanrat) [inline sort_sorttype.go:147]: I don't like using break labels here. It makes it hard to understand and debug.
Could this be re-written with loops/functions instead?

File: sort_sorttype.go [hunk 4]
```diff
@@ -190,13 +201,14 @@ func (s *SortTypeSorter) sortChunks() error {
...
 	scratch := make([]byte, binary.MaxVarintLen64)
+	var raw []byte
 	for {
 		select {
 		case b, more := <-s.saveChunkChan:
 			if more {
 				for _, d := range b.data {
 					// binary encoding for size
-					raw := d.ToBytes()
+					raw = d.ToBytes(raw[:0])
```
Comment (lanrat) [inline sort_sorttype.go:211]: Why is this being done? it seems inefficient and confusing.
This should be 2 different variables to be more clear, although I'm still not convinced its a good way to do this.

File: types.go [hunk 5]
```diff
@@ -4,7 +4,15 @@ import "context"
 
 // SortType defines the interface required by the extsort library to be able to sort the items
 type SortType interface {
-	ToBytes() []byte // ToBytes used for marshaling with gob
+	ToBytes(buf []byte) []byte
+}
+
+type Sizer interface {
```
Comment (lanrat) [inline types.go:10]: Please add a doc comment explaining the purpose of this interface

File: types.go (hunk 1 above)
Comment (lanrat) [inline types.go:14]: Please add a doc comment explaining the purpose of this interface

Comment (lanrat) [reply types.go:14]: Yes. I don't understand why they were added.

File: types.go [hunk 6]
```diff
@@ -4,7 +4,15 @@ import "context"
 
 // SortType defines the interface required by the extsort library to be able to sort the items
 type SortType interface {
-	ToBytes() []byte // ToBytes used for marshaling with gob
+	ToBytes(buf []byte) []byte
```
Comment (lanrat) [inline types.go:7]: The doc comment here should be updated to explain the purpose of buf.

File: sort_sorttype.go [hunk 7]
```diff
@@ -19,10 +19,9 @@ type chunk struct {
 	less CompareLessFunc
 }
 
-func newChunk(size int, lessFunc CompareLessFunc) *chunk {
```
Comment (lanrat) [inline sort_sorttype.go:22]: If you happen to know the chunk size in advance this offers a huge performance gain by not resizing the array continuously. You don't need to always fill the array and the chunk will still work.

File: sort_sorttype.go [hunk 8]
```diff
@@ -19,10 +19,9 @@ type chunk struct {
...
+func newChunk(lessFunc CompareLessFunc) *chunk {
 	c := new(chunk)
 	c.less = lessFunc
```
Comment (lanrat) [inline sort_sorttype.go:24]: If `c.data` it not initialized it could cause a panic.

**Review by lanrat — CHANGES_REQUESTED** [review]
I still don't understand the purpose of the chunk size change.
And a lot of the changes in this CL will reduce performance.

Can you better explain the use-case that was the motivation for this change?

**Comment by lanrat** [comment]
Closing this PR since the above questions remain unanswered.
If you are still interested in getting this merged please reply and I'll re-open.

**Pull request closed without merging**